File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        int x = 0;
        int y = 0;
        int dx = 0;
        int dy = 1;
        vector<int> res;

        for(int i = 0; i<rows * cols; i++){
            res.push_back(matrix[x][y]);
            matrix[x][y] = -101;
            if(!(x+dx >= 0 && x+dx < rows && y+dy >= 0 && y+dy < cols) || matrix[x+dx][y+dy]==-101){
                int temp = dy;
                dy = -dx;
                dx = temp;
            }

            x += dx;
            y += dy;
        }

        return res;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        


        res = []

        row = len(matrix)
        col = len(matrix[0])

        rStep = 0
        cStep = 1
        x, y = 0, 0
        
        for i in range (row*col):

            res.append(matrix[x][y])
            matrix[x][y] = -101
            
            if (not (x+rStep >= 0 and x+rStep < row and y+cStep >= 0 and y+cStep < col) or matrix[x+rStep][y+cStep] == -101):
                
                temp = cStep
                cStep = -rStep
                rStep = temp

            x += rStep
            y += cStep

        return res